### ISSUE TITLE
[x86/Linux] 16-byte stack alignment violation in JIT code

### ISSUE THREAD

**Issue by parjong** [dotnet/coreclr#8590]
I currently attempt to run the following simple code on x86/Linux port:
```
using System;

namespace ConsoleApplication {
  public class Program {
    public static void Main(string[] args) {
      int x = 3;
      Console.WriteLine("Hello World!");
      Show(x);
      Throw();
    }

    public static void Show(int x) {
      Console.WriteLine("Nice to meet you! " + x.ToString());
    }

    public static void Throw() {
      throw new System.ArgumentException();
    }
  }
}
```
As I working on x86/Linux port with https://github.com/parjong/coreclr/commit/1af98dca5a34aa29d0cd0b760a2b174eb50342c5 (which disables necessary exception handling macros), I expect to see some errors such as 'uncaught exception'.

Unfortunately, the current x86/Linux port ends up with segmentation fault as follows:
```
Hello World!
Nice to meet you!
Segmentation fault (core dumped)
```

**Comment by parjong** [comment]
\CC @seanshpark

**Comment by parjong** [comment]
Here is stack trace when segmentation fault occurs:
```
Program received signal SIGSEGV, Segmentation fault.
SystemDomain::GetCallersModule (stackMark=0xffffbae0, ppAppDomain=0x0) at /home/parjong/projects/dotnet/coreclr/src/vm/appdomain.cpp:4248
4248    /home/parjong/projects/dotnet/coreclr/src/vm/appdomain.cpp: No such file or directory.
(gdb) x/1i $pc
=> 0xf71bdb6e <SystemDomain::GetCallersModule(StackCrawlMark*, AppDomain**)+62>:        movaps %xmm0,-0x38(%ebp)
(gdb) p/x $ebp
$1 = 0xffffb98c
(gdb) bt
#0  SystemDomain::GetCallersModule (stackMark=0xffffbae0, ppAppDomain=0x0) at /home/parjong/projects/dotnet/coreclr/src/vm/appdomain.cpp:4248
#1  0xf71bdcf6 in SystemDomain::GetCallersAssembly (stackMark=0xffffbae0, ppAppDomain=0x0) at /home/parjong/projects/dotnet/coreclr/src/vm/appdomain.cpp:4272
#2  0xf7339006 in AssemblyNative::GetExecutingAssembly (stackMark=..., retAssembly=...) at /home/parjong/projects/dotnet/coreclr/src/vm/assemblynative.cpp:2161
#3  0xf0232813 in ?? ()
#4  0xf023263c in ?? ()
...
```
``movaps`` instruction requires 16-byte alignment, but the memory operand does not satisfy this requirement.

**Comment by parjong** [comment]
It seems that the issues is inside JITted code. While digging this issue, I found that 16-byte alignment is broken at the following frame:
```
(gdb) bt
...
#20 0xf2727166 in ?? ()
#21 0xf2722478 in ?? ()
#22 0xf71b0787 in CallDescrWorkerInternal () at /home/parjong/projects/dotnet/coreclr/src/vm/i386/asmhelpers.S:444
(gdb) info frame 21
Stack frame at 0xffffc530:
 eip = 0xf2722478; saved eip = 0xf71b0787
 called by frame at 0xffffc540, caller of frame at 0xffffc524
 Arglist at 0xffffc528, args:
 Locals at 0xffffc528, Previous frame's sp is 0xffffc530
 Saved registers:
  ebp at 0xffffc528, eip at 0xffffc52c
(gdb) info frame 20
Stack frame at 0xffffc524:
 eip = 0xf2727166; saved eip = 0xf2722478
 called by frame at 0xffffc530, caller of frame at 0xffffc518
 Arglist at 0xffffc51c, args:
 Locals at 0xffffc51c, Previous frame's sp is 0xffffc524
 Saved registers:
  ebp at 0xffffc51c, eip at 0xffffc520
```
(As I run this code with PR #8587, ``CallDescrWorkerInternal`` satisfies 16-byte stack alignment requirement)

**Comment by parjong** [comment]
@dotnet/jit-contrib @janvorli @jkotas IMHO, there will be two approaches to resolve this issue
 - Fix JIT (as in #8587), or
 - Fix Clang options (as in #8591).
Could you let me know which approach will be better to resolve this issue?

**Comment by jkotas** [comment]
Fixing the JIT would be the right thing to do.

The underlying platform libraries do assume the 16-byte alignment and it is not feasible to recompile them with different compiler flag. If we kept the JITed code or the runtime code on the 4-byte alignment plan, there needs to be some kind of alignment thunks on transitions to platform code. It is pretty non-trivial to get them built and they make everything slower (we have tried that in the past).

**Comment by BruceForstall** [comment]
So Linux/x86 assumes 16 byte stack alignment, compared to Windows/x86 only requiring 4 byte stack alignment?

In the JIT, you can change `STACK_ALIGN` and friends in target.h (under some appropriate `#ifdef`), which will get you most of the way, though presumably there's some carefully coded prolog/epilog code that will need to change, such as in `Compiler::lvaAlignFrame()`.

**Comment by parjong** [comment]
@jkotas @BruceForstall Thanks you for comment. I'll try that approach.

**Comment by janvorli** [comment]
@BruceForstall this is from wikipedia on x86 calling conventions (description is for cdecl):
> In the context of the C programming language, function arguments are pushed on the stack in the reverse order. In Linux, GCC sets the de facto standard for calling conventions. Since GCC version 4.5, the stack must be aligned to a 16-byte boundary when calling a function (previous versions only required a 4-byte alignment.)

**Comment by seanshpark** [comment]
// I think it's better to add status here than in PR...
With c918315 patch, including spilled registers, almost it seems working execpt for this case.
With `gentogen01.exe` unit test case, `System.AppDomain:SetupDomain(bool,ref,ref,ref,ref):this` is compile to have a `push 0` in the middle before a call.
```
~
       3909         cmp      dword ptr [ecx], ecx
       FF15F4A26DB2 call     [System.AppDomainSetup:InternalSetApplicationTrust(ref):this]
       90           nop      
       6A00         push     0
       8B4DD4       mov      ecx, gword ptr [ebp-2CH]
       8B55CC       mov      edx, gword ptr [ebp-34H]
       FF15E4936DB2 call     [System.AppDomain:SetupFusionStore(ref,ref):this]
       90           nop      
       90           nop 
~
```
This also needs to be fixed in some way.

**Comment by seanshpark** [comment]
`FEATURE_FIXED_OUT_ARGS` seems to enable preallocate and not use `push`. But got some compile errors. While looking to the problems I wasn't sure... so,
1) Can enabling `FEATURE_FIXED_OUT_ARGS` a good idea? or it would be better to add additional push before pushing arguments if stack is misaligned?
2) Is this block valid in `src/jit/CMakeLists.txt` ? 
```
if (CLR_CMAKE_PLATFORM_ARCH_I386)
    add_subdirectory(legacyjit)
    if (NOT CLR_BUILD_JIT32)
        add_subdirectory(compatjit)
    endif ()
endif (CLR_CMAKE_PLATFORM_ARCH_I386)
```
this will include `legacyjit` and defines `LEGACY_BACKEND` that had some un-implements in `codegenlegacy.cpp"

**Comment by seanshpark** [comment]
@jkotas, @janvorli , could you please tell me difference of JITs named `protojit`, `legacyjit` and `compatjit` that are in `src/jit/CMakeLists.txt` file ? It seems CoreCLR builds and run simple helloworld program without them for current x86/Linux.

**Comment by pgavlin** [comment]
> could you please tell me difference of JITs named protojit, legacyjit and compatjit that are in src/jit/CMakeLists.txt file ?

- protojit is a separate build of RyuJIT that can be used as an alt JIT (i.e. a JIT that is loaded alongside the usual JIT for purposes of testing/experimentation and platform bring-up)
- legacyjit is a build of the JIT that uses the legacy backend for code generation, and is used as an alt JIT
- compatjit is similar to legacyjit in that is uses the legacy backend, but it is not built as an alt JIT

Regarding the block in `src/jit/CMakeLists.txt` that you mentioned above, you can change that condition to `if (CLR_CMAKE_PLATFORM_ARCH_I386 AND WIN32)`--we shouldn't need these builds for x86/Linux.

**Comment by seanshpark** [comment]
@pgavlin , thanks a lot!

**Comment by seanshpark** [comment]
@jkotas , @pgavlin , could you help this problem?
I'm trying to make it work with `FEATURE_FIXED_OUT_ARGS` enabled.
Before the change, assembly generates with `push` comamnd, it generates like for example,
```
       8B4DEC       mov      ecx, dword ptr [ebp-14H]
       51           push     ecx
       8B4DE8       mov      ecx, dword ptr [ebp-18H]
       51           push     ecx
       8B4DD8       mov      ecx, gword ptr [ebp-28H]
       8B55D4       mov      edx, gword ptr [ebp-2CH]
       E83A27C701   call     System.String:LastIndexOfAny(ref,int,int):int:this
```
and with `FEATURE_FIXED_OUT_ARGS` to 1, and some other changes, now it uses `mov` command,
```
       8B4DEC       mov      ecx, dword ptr [ebp-14H]
       890C24       mov      dword ptr [esp], ecx
       8B4DE8       mov      ecx, dword ptr [ebp-18H]
       894C2404     mov      dword ptr [esp+04H], ecx
       8B4DD8       mov      ecx, gword ptr [ebp-28H]
       8B55D4       mov      edx, gword ptr [ebp-2CH]
       E80D374702   call     System.String:LastIndexOfAny(ref,int,int):int:this
```
code order with variables are same but the arguments in the stack are reversed.
Could you help where I should look at to fix this so that it would be passed as it should be?

**Comment by BruceForstall** [comment]
@seanshpark Just to understand the requirements: at what points is it required that the stack is 16 byte aligned? Is it that on entry to a function the stack was 16 byte aligned before the 'call' instruction pushed the 4-byte return address? So the callee knows to `sub esp,12` to get back to 16 byte alignment? In your example above of pushing 8 bytes of arguments, would an alternative be to push 8 bytes of zeros first? 

You'll also have to change the code that handles localloc to round up to 16 (although it looks like genLclHeap() already uses `STACK_ALIGN` so this might work already.)

You might be looking for targetx86.cpp:
```
const Target::ArgOrder Target::g_tgtArgOrder = ARG_ORDER_L2R;
```
and add an `#ifdef` to change it to R2L. Not sure, though.

I'm worried there might be other implications of changing x86 to use `FEATURE_FIXED_OUT_ARGS` that would be undesirable.

**Comment by seanshpark** [comment]
@BruceForstall, thank you for the reply!

> at what points is it required that the stack is 16 byte aligned?

I understand that before calling the function stack should be aligned to 16 byte as you mentioned with additonal questions and I also think so.

> In your example above of pushing 8 bytes of arguments, would an alternative be to push 8 bytes of zeros first?

Yes, I think this would do the align. But not sure this is a good approach. I've tried some but that was not enough and didn't run as expected. I think there are lots of knowledge to catch up.

`ARG_ORDER_L2R`. This is great. I'll try! 

> I'm worried there might be other implications of changing x86 to use FEATURE_FIXED_OUT_ARGS that would be undesirable.

I'm currently working on `FEATURE_FIXED_OUT_ARGS` enabled to match with other architectures on Linux. I've seen there are places where this feature is not handled with x86. I'll add this work in progress changes to the PR after some cleaning before it goes to some wrong direction.

**Comment by seanshpark** [comment]
Using `ARG_ORDER_R2L` turns out that input arguments are assigned to reverse order and there seems to be other places to touch, like stub functions. And doesn't seem to affect parameter order when calling to another function which I'm looking for.

**Comment by seanshpark** [comment]
I've updated #8849 with adding ESP adjustment before pushing arguments and after the call.
As this is my first changes to the JIT, it may look strange. I'm ready for any comments :)
I didn't want to change `GenTreePutArgStk` structure so used `slotNum` to hold alinment padding in the first argument. If adding a member is ok then I will change.

I wanted to use `FEATURE_FIXED_OUT_ARGS` but couldn't find the problem I've encountered. After this change is merged (I hope) I want to try again after major unit tests are cleared.
